src/main/java/io/wispforest/lavender/structure/StructureTemplate.java

```diff
@@ -163,224 +287,330 @@ public int countValidStates(World world, BlockPos anchor, BlockRotation rotation
+     * to be used when placing in the world
+     */
+    public Vec3i anchor() {
+        return this.anchor;
     }
 
-    // --- parsing ---
+    // --- iteration ---
 
-    @SuppressWarnings({"rawtypes", "unchecked"})
-    public static StructureTemplate parse(Identifier resourceId, JsonObject json) {
-        var keyObject = JsonHelper.getObject(json, "keys");
-        var keys = new Char2ObjectOpenHashMap<BlockStatePredicate>();
-        Vec3i anchor = null;
+    public void forEachPredicate(BiConsumer<BlockPos, BlockStatePredicate> action) {
+        this.forEachPredicate(action, BlockRotation.NONE);
+    }
 
-        for (var entry : keyObject.entrySet()) {
-            char key;
-            if (entry.getKey().length() == 1) {
-                key = entry.getKey().charAt(0);
-                if (key == '#') {
-                    throw new JsonParseException("Key '#' is reserved for 'anchor' declarations");
-                }
+    /**
+     * Execute {@code action} for every predicate in this structure template,
+     * rotated on the y-axis by {@code rotation}
+     */
+    public void forEachPredicate(BiConsumer<BlockPos, BlockStatePredicate> action, BlockRotation rotation) {
+        var mutable = new BlockPos.Mutable();
 
-            } else if (entry.getKey().equals("anchor")) {
-                key = '#';
-            } else {
-                continue;
-            }
+        for (int x = 0; x < this.predicates.length; x++) {
+            for (int y = 0; y < this.predicates[x].length; y++) {
+                for (int z = 0; z < this.predicates[x][y].length; z++) {
 
-            try {
-                var result = BlockArgumentParser.blockOrTag(Registries.BLOCK.getReadOnlyWrapper(), entry.getValue().getAsString(), false);
-                if (result.left().isPresent()) {
-                    var predicate = result.left().get();
+                    switch (rotation) {
+                        case CLOCKWISE_90 -> mutable.set(this.zSize - z - 1, y, x);
+                        case COUNTERCLOCKWISE_90 -> mutable.set(z, y, this.xSize - x - 1);
+                        case CLOCKWISE_180 -> mutable.set(this.xSize - x - 1, y, this.zSize - z - 1);
+                        default -> mutable.set(x, y, z);
+                    }
 
-                    keys.put(key, new BlockStatePredicate() {
-                        @Override
-                        public BlockState preview() {
-                            return predicate.blockState();
-                        }
+                    action.accept(mutable, this.predicates[x][y][z]);
+                }
+            }
+        }
+    }
 
-                        @Override
-                        public Result test(BlockState state) {
-                            if (state.getBlock() != predicate.blockState().getBlock()) return Result.NO_MATCH;
+    @NotNull
+    @Override
+    public Iterator<Pair<BlockPos, BlockStatePredicate>> iterator() {
```

Comment:
What precisely is the use case for this over the iteration methods? Having the entire iterator implementation below (which once again lays its fields out in a manner inconsistent with the rest of the codebase) seems like code I'd rather not have to maintain if there's not a good application

Reply:
my specific usecase was:

iterate over all the blocks in a structure, *exiting immediately* when it finds the first block that matches some conditions.

the reason `forEachPredicate` wasn't appropriate for me, was it would not exit after the first match. what I wanted was something that a `Stream` is perfect for, for which implementing `Iterable` was needed.

however, with `predicates` being exposed, I can implement that in my own codebase. but, for posterity I contributed both changes.

Reply:
Leaving this as unresolved while I wait for you to reply.